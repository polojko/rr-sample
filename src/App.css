.App {
  text-align: center;
  max-width: 800px;
  margin: 0 auto;
}

.App-logo {
  max-width: 90vw;
}

/* @media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
} */

.App-header, .App-body {
  background-color: #282c34;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}
.App-body {
  min-height: 60vh;
}

.App-link {
  color: #61dafb;
}


.loader {
  margin: 10vh 0;
  width: 100%;
}
.cards {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding: 0 0 3vh;
}
.card {
  display: flex;
  flex-direction: column;
  width: 220px;
  padding: 10px;
  margin: 10px;
  box-shadow: 0 10px 16px 0 rgba(0,0,0,0.2),0 6px 20px 0 rgba(0,0,0,0.19);
  background: rgb(236,215,151);
  background: linear-gradient(90deg, rgba(236,215,151,1) 0%, rgba(156,143,106,1) 30%, rgba(233,228,211,1) 100%);
}
.card:hover {
  transition: all 0.5s 0s ease;
  transform: scale(1.05,1.05);
}

.card-name {
  font-weight: bold;
  font-size: 1.1rem;
  padding-bottom: 1vh;
}
.card-text, .other-fields {
  font-size: .8rem
}

.select-card {
  padding: 1vh;
  font-size: 1rem;
  margin-top: auto;
}
.select-card input[type='checkbox'] {
  vertical-align: middle;
}

.card img {
  width: 100%;
}

.footer {
  display: block;
  clear: both;
  margin: auto;
  padding: 6px;
  text-align: center;
  position: fixed;
  left: 0;
  width: 100%;
  background-color: #fff;
  z-index: 1;
  bottom: 0;
}

.search {
  padding: 3vw;
}
.scroll-to-top {
  position: fixed;
  bottom: 30px;
  right: 10px;
  z-index: 1;
  cursor: pointer;
}

.search-message {
  padding-top: 1vh;
  color: crimson;
}

@media (max-width: 514px) {
  .scroll-to-top {
    bottom: 60px;

  }
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
