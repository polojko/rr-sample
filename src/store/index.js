import { createStore } from 'redux'

const initialState = {
  cards: [],
  pageSize: 20,
  page: 1,
  searchTerm: ''
}

export const LOAD_CARDS = 'LOAD_CARDS'
export const DO_SEARCH = 'DO_SEARCH'

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case LOAD_CARDS:
      return {
        ...state,
        page: ++state.page,
        cards: state.cards.concat(action.payload)
      }
    case DO_SEARCH:
        return {
            ...state,
            cards: [],
            page: 1,
            searchTerm: action.payload
            
        }
    default:
      return state
  }
}

const store = createStore(
  reducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

export default store