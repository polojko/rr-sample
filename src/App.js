import React, { useState, useEffect } from 'react'
import './App.css';
import { connect } from 'react-redux';
import _ from 'lodash';
import InfiniteScroll from "react-infinite-scroll-component";
import { Loader, NoData, Footer, Error, ScrollToTop } from './components/Helpers';
import SearchBar from "material-ui-search-bar";

function App(props) {

  const { cards, page, pageSize } = props;
  const [searchTerm, setSearchTerm] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const [showScroll, setShowScroll] = useState(false);
  const [error, setError] = useState('');

  const loadCards = (name, type) => {
    setIsLoading(true); // resets isLoading back to true
    let searchTerm = name ? name : '';
    let pageCount = (type && type === 'search') ? 1 : page;
    setTimeout(() => {
      fetch('https://api.elderscrollslegends.io/v1/cards?name=' + searchTerm + '&pageSize=' + pageSize + '&page=' + pageCount) // useSwr to cache this - yugo-music
        .then(response => response.json())
        .then(json => {
          console.log('json: ', json);
          setIsLoading(false);
          props.dispatch({
            type: 'LOAD_CARDS',
            payload: json.cards
          })
        })
        .catch(function (error) {
          setError(error);
          console.log(error);
        });
    }, 2000) // timeout to see loader

  }

  const doSearch = (input) => {
    props.dispatch({
      type: 'DO_SEARCH',
      payload: input
    })
    loadCards(input, 'search')
  }

  const loadingMessage = (isLoading, error) => {
    if (error) {
      return <Error />
    } else if (isLoading) {
      return <Loader />
    } else {
      return <NoData />
    }
  }

  const showScrollToTop = () => {
    console.log('scroll called!!');
    if (window.pageYOffset > 400) {
      setShowScroll(true)
    } else if (window.pageYOffset <= 400) {
      setShowScroll(false)
    }
  }

  useEffect(() => {
    loadCards();
    window.addEventListener('scroll', _.debounce(showScrollToTop,500));
  }, []) // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <div className="App">
      <header className="App-header">
        <img src='/logo.png' className='App-logo' alt='Elder Scrolls Legends' />
      </header>
      <div className='search'>
        <SearchBar
          value={searchTerm}
          onChange={(newValue) => setSearchTerm(newValue)}
          onRequestSearch={() => doSearch(searchTerm, 'search')}
          onCancelSearch={() => doSearch(setSearchTerm(''), 'search')} // clears cards, and refetches them
        />
      </div>
      <div className='cards-container'>
        <InfiniteScroll
          dataLength={cards.length}
          next={() => loadCards(searchTerm)}
          hasMore={true}
          loader={loadingMessage(isLoading, error)}
          className='cards'
          scrollThreshold='100px'
        >
          {cards.map((card, index) => (
            <div key={card.id} className='card'>
              <img src={card.imageUrl} alt={card.name} />
              <div className='card-name'>{card.name}</div>
              <div className='card-text'>{card.text}</div>
              <div className='other-fields'>
                {card.set.name && <span><b>Set:</b> {card.set.name} </span>}
                {card.type && <span><b>Type:</b> {card.type}</span>}
              </div>
            </div>
          ))}
        </InfiniteScroll>
      </div>
      {showScroll && <ScrollToTop />}
      <Footer />
    </div>
  )
}

const mapStateToProps = state => {
  return {
    cards: state.cards,
    page: state.page,
    pageSize: state.pageSize,
    searchTerm: state.searchTerm
  }
}

const mapDispatchToProps = dispatch => {
  return {
    dispatch
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)