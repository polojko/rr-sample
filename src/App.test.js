import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import App from './App';
import { Provider } from 'react-redux';
import store from './store/index';


afterEach(cleanup);

it('App component renders correctly', () => {
  const storeData = {

  }
  const app = renderer
      .create(<Provider store={store}>
        <App />
      </Provider>)
      .toJSON();
  expect(app).toMatchSnapshot();
});
