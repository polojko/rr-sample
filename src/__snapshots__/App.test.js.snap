// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> App component renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="Elder Scrolls Legends"
      className="App-logo"
      src="/logo.png"
    />
  </header>
  <div
    className="search"
  >
    <div
      className="MuiPaper-root ForwardRef-root-1 MuiPaper-elevation1 MuiPaper-rounded"
      style={null}
    >
      <div
        className="ForwardRef-searchContainer-7"
      >
        <div
          className="MuiInputBase-root MuiInput-root ForwardRef-input-6 MuiInputBase-fullWidth MuiInput-fullWidth"
          onClick={[Function]}
        >
          <input
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="Search"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiIconButton-root ForwardRef-iconButton-2 ForwardRef-searchIconButton-4"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root ForwardRef-icon-5"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiIconButton-root ForwardRef-iconButton-2 ForwardRef-iconButtonHidden-3"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root ForwardRef-icon-5"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      className="search-message"
    >
      
    </div>
  </div>
  <div
    className="cards-container"
  >
    <div
      className="infinite-scroll-component__outerdiv"
      style={Object {}}
    >
      <div
        className="infinite-scroll-component cards"
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <div
          className="loader"
        >
          <div
            className="MuiCircularProgress-root MuiCircularProgress-indeterminate"
            role="progressbar"
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          >
            <svg
              className="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx={44}
                cy={44}
                fill="none"
                r={20.2}
                strokeWidth={3.6}
                style={Object {}}
              />
            </svg>
          </div>
          <p>
            Loading...
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    © 
    2020
     The Elder Scrolls Legends™. All rights reserved
  </div>
</div>
`;
